Implement proxy redirection logic in eBPF

joestringer opened the pull request:
### Summary

Introduce a new implementation of the proxy redirect handling in the eBPF datapath using a combination of socket lookups and socket assign (+ packet redirects where necessary to ensure that socket assign occurs at TC ingress).

This functionality is disabled by default in the daemon, but enabled in new v1.9 deployments via Helm. When enabled on Linux 5.7 or later, it removes use of the `-j TPROXY` target, and removes Cilium's usage of the upper 16 bits of the mark for the purpose of sharing the tproxy target port with iptables rules.

### API impact

Three new monitor error codes are added for cases where the redirect cannot be applied, meaning that traffic subject to L7 policy will be dropped if there is a configuration or lookup issue:
*  `DROP_PROXY_LOOKUP_FAILED`
   * Lookup for the target proxy socket failed. This will trigger drop when the datapath cannot find any establish proxy connection for this tuple AND the datapath cannot find a TPROXY listening socket corresponding to the redirect port in the eBPF policy map. If the proxy is not available for any reason then this could be a reason to see this error.
* `DROP_PROXY_SET_FAILED`
   * Failed to assign the socket to the packet context. Typically programmer error but may trigger if the socket which was found previously is unavailable or has incompatible options set (eg `SO_REUSEPORT`). In the event of this error, the default monitor logs will additionally provide an errno to determine which subfailure occurred.
* `DROP_PROXY_UNKNOWN_PROTO`
   * Safety catchall in case we attempt to redirect traffic other than TCP/UDP to a socket. As of Linux 5.7, no other L4 protocols are supported for tproxy redirection.

### Implementation

Previously, in the eBPF datapath during policy apply and redirect processing, we have relied purely on the tproxy port which is available in the policymap to prepare the packet for TPROXY redirection. Mechanically this path involved performing a policy lookup using the src/dst identity, then in the policymap entry, using non-zero redirect port as a signal for redirect. This port would then be set into the `ctx->mark` along with some bits to identify this traffic as "needs proxy redirection". Then the packet would be passed up the stack towards an iptables PREROUTING rule which would match on the mark then perform a socket lookup and assign the socket to the skb.

The first step of the new implementation is to ensure that the relevant forwarding logic has access to the full packet tuple. There is one fairly significant condition for applying the socket to the skb: It must be done at TC ingress. In many cases, Cilium already applies all logic at TC ingress eBPF programs so this isn't too difficult. For `endpoint-routes` mode however we apply pod ingress policy at the TC egress hook point, and therefore must first hairpin to another device, then extract the tuple & perform the socket lookups to assign the socket to the skb.

With this in place, the actual lookup & assign is relatively straightforward. First we must look for any established connections corresponding to the tuple in the local network namespace.These represent existing proxy connections that were already established. If we find the socket here, we assign it to the skb & pass up the stack. If we don't find it, then we zero out the tuple and set the TPROXY redirect port from the policy into the tuple and perform a second lookup for listen sockets. This should provide the parent listener socket associated with the proxy which we can assign to establish a new connection in the proxy.

For now, this PR introduces dedicated tuple extraction logic in `bpf_host.c` for the endpoint routes mode case rather than sharing it with host firewall, primarily because it predates host firewall but also because the tuple extraction needs to be performed if eBPF tproxy is enabled but host firewall is not. A future cleanup could look at shifting tuple extraction into common code which would be enabled by either path and shared. Further optimizations could be explored to perhaps avoid further packet touching in this prog by sharing some of that metadata from the previous program that was run at TC egress of the endpoint subject to L7 policy.

This PR replaces the per-redirect iptables rule logic but not the static `-m socket` logic as this requires further exploration to determine the simplest path forward. Fortunately, the newly added connectivity checks for proxy paths makes it easy to reproduce failures on these paths to allow investigation and experimentation.

### Testing

The recently-added `examples/kubernetes/connectivity-check-proxy.yaml` is now added to the main CI for testing various proxy paths to prevent regressions. Additionally I have manually deployed this YAML for debugging in a GKE environment (default instructions, `endpoint-routes` mode).

The `BPF checks / checkpatch (pull_request)` GitHub action job doesn't seem to like my use of macros to avoid duplicating IPv4/IPv6 implementations, but I've looked through the complaints and AFAICT they're all safe to ignore.

### Tasks
* [x] Daemon logic to override enable/disable
  * Iptables compatibility mode that still configures rules for compatibility with existing rulesets (eg kube-proxy)
* [x] IPv6 support
* [x] Endpoint-routes mode validation
* [x] CI tests to validate the new paths
* [x] Issue description for reviewers :-)

Deferred to follow up PRs
* [ ] Replace `-m socket` rules - I looked briefly into this, should be doable in one of a couple of ways but it's more involved so I figured it's better to get the basics in first.
* [x] This PR touches the issue https://github.com/cilium/cilium/issues/12288 but does not resolve it.

I marked this 'release-note/major' with the (ambitious?) goal of it being listed with other PRs as part of iptables removal for v1.9, but ideally it should have no user-facing impact :-) Happy to move it to /minor if we think this is not v1.9 release higlight worthy.

Related: #9921

joestringer committed:
bpf: Add socket lookup,assign helpers

Define these helpers for use in upcoming BPF TPROXY commits.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
monitor: Add notifications for socket failures

Add drop and debug notifications that will be used by upcoming commits
to emit informational events from the datapath when BPF TPROXY is
enabled.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Pass tuple into ctx_redirect_to_proxy

In preparation for implementing BPF TPROXY in ctx_redirect_to_proxy(),
pass the tuple into this function from all of its callers. In this
commit it will remain unused but in future commits this will begin to be
used in particular cases to perform a socket lookup in the callee.

When endpoint-routes mode is enabled, we're executing on TC egress which
can't make use of the tuple yet so we won't extract the tuple out at
this point (pass NULL). The "hairpin" redirect will redirect this back
to a device where we can perform the tuple extraction & socket lookup.

NOTE: At this point when passing the tuple in, I am making the
assumption that it is no longer needed at these call points so
ctx_redirect_to_proxy() is welcome to thrash it and use it as scratch
buffer as it searches for the relevant sockets. As such it is
deliberately not marked "const". If this assumption is invalid or
reviewers have suggestions how to alleviate this assumption, please
comment in the review.

If you're reading this in the future and have hit an issue due to
rewrites of the tuple in this function, then.... sorry? :-)

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Implement BPF TPROXY for traffic to the host

In cases where the traffic is being processed in a BPF program at TC
ingress, ie the traffic is already heading towards the host, we can
implement the socket lookup & assign logic directly in the proxy
redirect logic here. This applies, for instance, to traffic arriving
from an endpoint on the lxc device, or from a tunnel (via tail call to
the endpoint policy program).

In these cases, we attempt to first locate any full sockets
associated with the tuple, and if the connection is already established,
assign the corresponding socket to the skb to notify the stack to pass
the traffic into the socket. If the full socket cannot be found, then we
fall back to listen sockets for the proxy port and assign that one.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Use CB_PROXY_MAGIC to store proxy meta magic

Use the define here to improve readability and locating the code that
changes specific offsets in the context metadata (eg skb->cb). The value
is the same, this is a purely cosmetic change.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Fix up proxy debug monitor message for hairpin

When hairpinning, it's useful to have two datapoints for the proxy
redirect: One in the original program and one in the program that
actually prepares the context to send the packet to the proxy. The
ctx_redirect_to_proxy_hairpin one should occur in the original program,
so change it to DBG_CAPTURE_PROXY_PRE. The corresponding
DBG_CAPTURE_PROXY_POST should be handled in the "to-host" program later.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Add --enable-bpf-tproxy

This flag supports manually enabling or disabling the BPF-based tproxy
functionality as supported on Linux 5.7 or later. If it is enabled, then
the iptables rules which perform tproxy action will be disabled and the
equivalent functionality will be enabled in BPF in the datapath. Some
rules such as ones that configure the stack to ignore proxy flows and
avoid connection tracking them via iptables are left in place in case
the user still has iptables enabled in their environment.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Add proxy datapaths test

Add a test which deploys connectivity-check-proxy.yaml from the
examples/kubernetes/connectivity-check/ dirs, to validate the various
egress/ingress/dual proxy cases.

Related: #12714

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
install: Add support to configure BPF tproxy

Default it to enabled for new v1.9 installs, v1.8 upgrades will disable.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Implement BPF TPROXY for traffic from the host

For stack packet receive ~reasons~, the ability to assign a socket to
the context is only supported at TC ingress. As such, when for instance
the host itself is sending traffic to endpoints on the same node, or
when the egress proxy is sending traffic towards an endpoint that also
has ingress L7 policy applied, we need to redirect the packet towards an
ingress TC program before assigning the socket to the context.

The process of redirecting the traffic towards an ingress qdisc was
already present as this is necessary to actually present the traffic to
the stack to send to the proxy; this commit extends that logic to
perform the same logic as the previous commit for socket assign from the
host ingress program.

In the case of the host ingress program, we don't have the luxury of
having already extracted the packet tuple for performing socket lookup,
so in this case we are required to perform this extraction again. (*)

(*) This was written pre-hostfw so there may be some room to perform
better sharing of the implementation for this case now.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on install/kubernetes/cilium/charts/config/templates/configmap.yaml:
```diff
@@ -29,6 +30,11 @@
 {{- $defaultBpfCtAnyMax = 0 -}}
 {{- end -}}
 
+{{- /* Default values when 1.9 was initially deployed */ -}}
+{{- if semverCompare ">=1.9" (default "1.9" .Values.upgradeCompatibility) -}}
+{{- $defaultBpfTProxy = "true" -}}
+{{- end -}}
+
```
@tgraf PTAL, I'm not sure if this is how you intended v1.9 Helm compatibility extension to work.

aanm replied:
I think we also need to fix https://github.com/cilium/cilium/issues/12288

joestringer replied:
Github is hiding this thread at this point, but this is the remaining piece to figure out /cc @tgraf

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -146,7 +146,7 @@ static __always_inline int ipv6_l3_from_lxc(struct __ctx_buff *ctx,
 	/* Check it this is return traffic to an ingress proxy. */
```
Commit `bpf: Pass tuple into ctx_redirect_to_proxy` has a line with `squash! bpf: Pass tuple into ctx_redirect_to_proxy` at the end of the commit message. I'm assuming this isn't intentional.

joestringer replied:
Ah thanks, I'll clip that out.

pchaigno commented on bpf/lib/proxy.h:
```diff
@@ -44,6 +187,8 @@ __ctx_redirect_to_proxy(struct __ctx_buff *ctx, void *tuple __maybe_unused,
 {
+	int result = CTX_ACT_OK;
+
 	ctx->mark = MARK_MAGIC_TO_PROXY | proxy_port << 16;
```
Is there a reason we still set the mark here even if TPROXY is enabled?

joestringer replied:
Yes. The `MARK_MAGIC_TO_PROXY` is still required for routing the packets locally, even if we assign the socket in this function at TC ingress.

In the case where this function is called at a TC egress hook point, it is additionally required to pass the port along to the subsequent eBPF program at TC ingress which will handle the socket lookup.

pchaigno replied:
I see. And `proxy_port` will always be zero in TC ingress case, no? Would it be worth writing the proxy port in the mark only in the TC egress case? I mostly want to avoid trashing mark bits if we don't need to.

joestringer replied:
I think it's a little more subtle than how you put it, but yes I believe that if we're at TC ingress we don't need to set the bits that we would otherwise dedicate to the proxy port. I'll look into another patch for this.

joestringer replied:
Proposed incremental:

```
diff --git a/bpf/lib/proxy.h b/bpf/lib/proxy.h
index 1efdc534ec83..cf7157bbc7d4 100644
--- a/bpf/lib/proxy.h
+++ b/bpf/lib/proxy.h
@@ -194,7 +194,12 @@ __ctx_redirect_to_proxy(struct __ctx_buff *ctx, void *tuple __maybe_unused,
 {
        int result = CTX_ACT_OK;
 
-       ctx->mark = MARK_MAGIC_TO_PROXY | proxy_port << 16;
+#ifdef ENABLE_TPROXY
+       if (!from_host)
+               ctx->mark |= MARK_MAGIC_TO_PROXY;
+       else
+#endif
+               ctx->mark = MARK_MAGIC_TO_PROXY | proxy_port << 16;
 
 #ifdef HOST_REDIRECT_TO_INGRESS
        cilium_dbg(ctx, DBG_CAPTURE_PROXY_PRE, proxy_port, 0);
```

joestringer committed:
dnsproxy: Disable SO_REUSEPORT with bpf tproxy

The upstream BPF TPROXY support doesn't support REUSEPORT sockets at
this time, so disable REUSEPORT if BPF TPROXY is enabled.

Signed-off-by: Joe Stringer <joe@cilium.io>

pchaigno commented on bpf/lib/proxy.h:
```diff
@@ -10,6 +10,149 @@
+	/* tuple's mismatched dport/sport strikes again! */			\
+	port = tuple->SK_FIELD.sport;						\
+	tuple->SK_FIELD.sport = tuple->SK_FIELD.dport;				\
+	tuple->SK_FIELD.dport = port;						\
```
Is this needed because `ct_lookup{4,6}` inversed the tuple? If that's it, could you add a comment in that sense?

joestringer replied:
I guess that's what my `tuple's mismatched dport/sport strikes again!` comment was attempting to convey, but evidently it's a bit cryptic ;)

joestringer replied:
Incremental:

```
diff --git a/bpf/lib/proxy.h b/bpf/lib/proxy.h
index 43aae77d13de..1efdc534ec83 100644
--- a/bpf/lib/proxy.h
+++ b/bpf/lib/proxy.h
@@ -115,7 +115,12 @@ NAME(struct __ctx_buff *ctx, CT_TUPLE_TYPE * ct_tuple, __be16 proxy_port)     \
        __u16 port;                                                             \
        int result;                                                             \
                                                                                \
-       /* tuple's mismatched dport/sport strikes again! */                     \
+       /* The provided 'ct_tuple' is in the internal Cilium format, which      \
+        * reverses the source/destination ports as compared with the actual    \
+        * packet contents. 'bpf_sock_tuple' in the eBPF API needs these to     \
+        * match normal packet ordering to successfully look up the             \
+        * corresponding socket. So, swap them here.                            \
+        */                                                                     \
        port = tuple->SK_FIELD.sport;                                           \
        tuple->SK_FIELD.sport = tuple->SK_FIELD.dport;                          \
        tuple->SK_FIELD.dport = port;
```

pchaigno commented on bpf/lib/proxy.h:
```diff
@@ -235,17 +235,107 @@ ctx_redirect_to_proxy6(struct __ctx_buff *ctx, void *tuple __maybe_unused,
+	 * TODO: Consider passing ports with meta/mark from		\
+	 * __ctx_redirect_to_proxy() to avoid touching packet here	\
+	 */								\
+	if (ctx_load_bytes(ctx, l4_off, &tuple->dport, 4) < 0)		\
```
I'm not sure I understand the comment. We're touching the packet in `ipv{4,6}_extract_tuple` anyway. Or do you mean we could avoid touching the L4 in particular?

joestringer replied:
Yeah, L4 is a little bit trickier just because you have to figure out the l4 offset so I figured maybe we can optimize that part out at least. But I don't know that it's actually worth it. I have no immediate intention to address this TODO, so perhaps I should just drop it.

pchaigno reviewed (commented):
I'm not sure I understand the ingress vs. egress subtleties of the proxy redirects, so I'll have to come back to that. LGTM on the first pass. I have a few comments below, mostly questions.

aanm reviewed (approved):
LGTM, only a small FYI that we need to address

christarazi commented on bpf/lib/proxy.h:
```diff
@@ -10,6 +10,149 @@
+#define CTX_REDIRECT_FN(NAME, CT_TUPLE_TYPE, SK_FIELD,				\
+			DBG_LOOKUP_CODE, DADDR_DBG, SADDR_DBG)			\
+static __always_inline int							\
+NAME(struct __ctx_buff *ctx, CT_TUPLE_TYPE * ct_tuple, __be16 proxy_port)	\
```
I see this is a macro definition, but the docstring doesn't match. It's instead the docstring of a usage of the macro (further down below). I think it would make sense to move the `ctx_redirect_to_proxy_ingress4 / 6` to where it's actually used and further document the macro. It's sort of confusing at first glance. If this was intentional though, I'm curious why :)

Also the spacing in the docstrings for the args are a bit messed up.

joestringer replied:
I wasn't planning on documenting the macro, so I figured it'd be more useful to provide examples and actual types of the declared functions for reference here to make it easier to reason about the logic defined by the macro. Perhaps if I moved the docstring inside the macro definition it would be more clear that it applies to the functions declared by the macro.

Also, I specifically wanted to include the function names explicitly here so that it's easier to grep and find this usage and know that this macro is defining the function to provide a hint to readers that "yes this is the piece you're looking for".

The spacing is not messed up if you open your editor using the standard settings for kernel dev (tabs=8wide), this is a known conflict with git (apparently including github) rendering of the lines.

joestringer replied:
I've queued up a change to move the docstrings inside the macros.

christarazi commented on test/k8sT/Policies.go:
```diff
@@ -1403,6 +1406,29 @@ var _ = Describe("K8sPolicyTest", func() {
+				kubectl.ApplyDefault(connectivityCheckYml).ExpectSuccess("cannot install connectivity-check")
+
+				err := kubectl.WaitforPods(helpers.DefaultNamespace, "", helpers.HelperTimeout)
+				ExpectWithOffset(1, err).Should(BeNil(), "connectivity-check pods are not ready after timeout")
```
`ExpectWithOffset` will show that the test failed at the *caller* of this `It()` block (which would be somewhere in ginkgo-ext or similar), IIRC. I think what you want is just an `Expect`, because that'll generate a stacktrace when the test fails, pointing at this specific line, which when a user looks at, it is clear that the error was with `WaitforPods`.

In general, I think if you are doing an expectation inside the `It()` block, you should use a simple `Expect`. An example of when you would use `ExpectWithOffset` is when you've defined a function that *performs* expectations. That's when you'd want to "offset" it. Does that make sense?

joestringer replied:
Good catch, I was copy/pasting this from a similar example elsewhere in the testsuite and didn't pay close attention. Will fix it up.

christarazi reviewed (approved):
LGTM; just a few comments / questions.

joestringer committed:
bpf: Avoid setting all bits on bpf stack ingress

When making use of eBPF TPROXY, avoid setting bits in the upper portions
of the mark which would previously be used in non-eBPF TPROXY for
determining the transparent socket port. These bits do not need to be
set in this case, so leaving them unset helps to better integrate with
other potential users of the mark.

Suggested-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Joe Stringer <joe@cilium.io>

errordeveloper reviewed (approved):
CLI and Helm changes LGTM.

borkmann commented on bpf/include/bpf/helpers.h:
```diff
@@ -84,4 +84,15 @@ static int BPF_FUNC(sock_hash_update, struct bpf_sock_ops *skops, void *map,
+				 __u64 netns, __u64 flags);
+static int BPF_FUNC(sk_release, struct bpf_sock *sk);
+static int BPF_FUNC(sk_assign, struct __sk_buff *skb, struct bpf_sock *sk,
+		    __u64 flags);
```
These should go into helpers_skb.h since they are all skb related.

borkmann commented on bpf/bpf_lxc.c:
```diff
@@ -905,6 +905,8 @@ ipv6_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, __u8 *reason,
 		send_trace_notify6(ctx, TRACE_TO_PROXY, src_label, SECLABEL, &orig_sip,
 				  0, ifindex, 0, monitor);
+		if (tuple_out)
+			*tuple_out = tuple;
```
nit: any reason to not use optimised memcpy we have?

joestringer replied:
No reason, makes sense to fix that up.

borkmann commented on bpf/lib/proxy.h:
```diff
@@ -10,6 +10,154 @@
+		  struct bpf_sock_tuple *tuple, __u32 len, bool established)
+{
+	int result = DROP_PROXY_LOOKUP_FAILED;
+	struct bpf_sock *sk = NULL;
```
nit: sk does not need assignment

borkmann commented on bpf/lib/proxy.h:
```diff
@@ -10,6 +10,154 @@
+		  bool established __maybe_unused)
+{
+	int result = DROP_PROXY_LOOKUP_FAILED;
+	struct bpf_sock *sk = NULL;
```
ditto
